linux.cc

```diff
@@ -386,9 +387,40 @@ static long sys_getcwd(char *buf, unsigned long size)
 // to Linux signature of this system call. The underlying ioctl function which we delegate to
 // is variadic and takes slightly different paremeters and therefore cannot be used directly
 // as other system call implementations can.
+#define __NCCS 19
+struct __kernel_termios {
+    tcflag_t c_iflag;
+    tcflag_t c_oflag;
+    tcflag_t c_cflag;
+    tcflag_t c_lflag;
+    cc_t c_line;
+    cc_t c_cc[__NCCS];
+    speed_t c_ispeed;
+    speed_t c_ospeed;
+};
```

Comment:
Please add a comment where you got this particular definition. 

Is this ktermios from /usr/include/asm-generic/termbits.h? (by the way, I have no idea  what are "termios" and "termios2" in the same header file...).

I looked in the glibc source code, and it assumes (sysdeps/unix/sysv/linux/kernel_termios.h) a slightly different definition.
```c
/* The following corresponds to the values from the Linux 2.1.20 kernel.  */

#define __KERNEL_NCCS 19

struct __kernel_termios
  {
    tcflag_t c_iflag;           /* input mode flags */
    tcflag_t c_oflag;           /* output mode flags */
    tcflag_t c_cflag;           /* control mode flags */
    tcflag_t c_lflag;           /* local mode flags */
    cc_t c_line;                /* line discipline */
    cc_t c_cc[__KERNEL_NCCS];   /* control characters */
  };
```
Without the speed members, but then glibc's source code sysdeps/unix/sysv/linux/tcgetattr.c assumes a c_ispeed and c_ospeed members in this structure. So I'm completely confused now.
Maybe worth checking what the actual kernel has.

Reply:
You are right about the `__kernel_termios` - we should make it like in `kernel_termios.h`. 

Regarding `c_ispeed` and `c_ospeed`, I think it is all OK given this code from glibc `tcgetattr.c`:
```c
      termios_p->c_line = k_termios.c_line;
#if _HAVE_STRUCT_TERMIOS_C_ISPEED
# if _HAVE_C_ISPEED
      termios_p->c_ispeed = k_termios.c_ispeed;
# else
      termios_p->c_ispeed = k_termios.c_cflag & (CBAUD | CBAUDEX);
# endif
#endif
```

The `_HAVE_STRUCT_TERMIOS_C_ISPEED` is 1 in the generic header but `_HAVE_C_ISPEED` is 0 so in actuality, glibc would take the information from `c_cflag` instead. The code accounts for `c_ispeed` that is present for mips and powerpc architectures.